### PR TITLE
Fixes #8834 feat(nimbus): Compute fields based on release date or start date

### LINKED ISSUE #8834: Use release date to calculate durations
The new release date field should be used to calculate the following fields:

* Launch month
* Proposed enrollment end date
* Proposed end date
* Computed enrollment days
* Computed enrollment end date
* Enrollment duration
* Computed duration days

!image-20230419-185320.png|width=1350,height=184!

See notes here: [https://docs.google.com/document/d/1nC_bcAQuGdo8T2ck3v-6HdT9olFffpFoct63CPzbU-w/edit](https://docs.google.com/document/d/1nC_bcAQuGdo8T2ck3v-6HdT9olFffpFoct63CPzbU-w/edit|smart-link)

┆Issue is synchronized with this [Jira Task](https://mozilla-hub.atlassian.net/browse/EXP-3488)
┆Attachments: <a href="https://mozilla-hub.atlassian.net/rest/api/2/attachment/content/37240">image-20230419-185320.png</a>

### PR DESCRIPTION
**Pull request by eliserichards**
Because

- If an experiment is first run, we might have a release date
- And the following fields on the model are calculated using start date:
   - Launch month
   - Proposed enrollment end date
   - Proposed end date
   - Computed enrollment days
   - Computed enrollment end date
   - Enrollment duration
   - Computed duration days

This commit

- Add `release_date` field, which checks if the experiment is first run (`is_first_run`) before checking if `proposed_release_date` exists
- Add `enrollment_start_date` field, which returns `proposed_release_date` if it exists on a first run experiment and `start_date` otherwise
- DOES NOT change the `start_date` calculations
- `start_date` will still be used in all calculations where the experiment is not first-run and has no proposed release date

### CODE REVIEW

File: experimenter/experimenter/experiments/models.py
Comment (yashikakhurana) [inline experimenter/experimenter/experiments/models.py:779]: good point, we always miss the clone part

File: experimenter/experimenter/experiments/tests/test_models.py
Comment (yashikakhurana) [inline experimenter/experimenter/experiments/tests/test_models.py:847]: why this assert? 🤔

File: experimenter/experimenter/experiments/tests/test_models.py
Comment (yashikakhurana) [inline experimenter/experimenter/experiments/tests/test_models.py:958]: don't we want to assert with something? 🤔

File: experimenter/experimenter/experiments/tests/test_models.py
Comment (yashikakhurana) [inline experimenter/experimenter/experiments/tests/test_models.py:1075]: do we need to pass `None`, I think by default it should be None

**Review by yashikakhurana — COMMENTED** [review]
In general, looks good, the only thing to reduce if I can see is to take out this common line `if self.is_first_run and self.proposed_release_date:`to have in a separate method, thoughts?

File: experimenter/experimenter/experiments/tests/test_models.py
Comment (eliserichards) [reply experimenter/experimenter/experiments/tests/test_models.py:847]: These tests are all copy/pasted from the existing tests to extend to testing `proposed_release_date` in addition to `start_date` :)

File: experimenter/experimenter/experiments/tests/test_models.py
Comment (eliserichards) [reply experimenter/experimenter/experiments/tests/test_models.py:958]: same ^ I'm happy to take these out, I was just following what was previously there

**Comment by eliserichards** [comment]
I added a helper function for checking `is_first_run` and `proposed_release_date`, but the problem is that the type checking still thinks it's possible for `proposed_release_date` to be `None` even if the function returns `True`. I tried a few things (like [TypeGuarding](https://peps.python.org/pep-0647/#typeguard-type) the function) but just landed on `assert`ing `proposed_release_date`. I'm not sure if this makes it better or worse lol. What do you think @yashikakhurana?

File: experimenter/experimenter/experiments/models.py
Comment (eliserichards) [inline experimenter/experimenter/experiments/models.py:449]: @yashikakhurana if I don't have the `assert` here, line 449 will error: 
```
error: "strftime" is not a known member of "None" (reportOptionalMemberAccess)
```

**Comment by eliserichards** [comment]
Ok, let me know what you think @yashikakhurana. It still requires the `assert`s, but the helper function will never return `None` now

File: experimenter/experimenter/experiments/models.py
Comment (yashikakhurana) [inline experimenter/experimenter/experiments/models.py:486]: same here

File: experimenter/experimenter/experiments/tests/test_models.py
Comment (yashikakhurana) [reply experimenter/experimenter/experiments/tests/test_models.py:847]: I think we should assert with some value as expected value

**Review by yashikakhurana — DISMISSED** [review]
In general looks good, you can remove assert from the non test cases code and also it would be helpful if we can assert with some value in test cases

File: experimenter/experimenter/experiments/models.py
Comment (eliserichards) [reply experimenter/experimenter/experiments/models.py:475]: You can't remove the asserts (or you need to replace them with self.proposed_release_date != None). The code does not know that it is not None, even with the helper function

Comment (jaredlockhart) [reply experimenter/experimenter/experiments/models.py:475]:
```py
if release_date := self..... is not None:
    ....
```

File: experimenter/experimenter/experiments/models.py
Comment (jaredlockhart) [inline experimenter/experimenter/experiments/models.py:445]:
```py
return bool(...)
```

Comment (jaredlockhart) [reply experimenter/experimenter/experiments/models.py:445]:
```py
def release_date(self):
    if self.is_first_run:
        return self.proposed_release_date
```

Comment (jaredlockhart) [reply experimenter/experimenter/experiments/models.py:445]:
```py
def enrollment_start_date(self):
    return self.release_date or self.start_date
```

File: experimenter/experimenter/experiments/models.py
Comment (eliserichards) [inline experimenter/experimenter/experiments/models.py:450]: These two fields replace the need for a helper function, and make it clear that:
1. `self.proposed_release_date` is dependent on `is_first_run`
2. The start date for enrollment will either be `release_date` or `start_date` in all places

**Review by brennie — APPROVED** [review]
LGTM but you may want to wait on review from someone more familiar with this than I am.

**Review by yashikakhurana — APPROVED** [review]
perfect 💯

File: experimenter/experimenter/experiments/tests/test_models.py
Comment (jaredlockhart) [inline experimenter/experimenter/experiments/tests/test_models.py:2076]: 💯 💯 💯 💯 💯 💯 💯

**Pull request merged**
